main/main.go

```diff
@@ -0,0 +1,298 @@
+		},
+		cli.BoolFlag{
+			Name:  "ignore-mac",
+			Usage: "ignore Message Authentication Code during decryption",
+		},
+		cli.StringFlag{
+			Name:  "unencrypted-suffix",
+			Usage: "override the unencrypted key suffix. default: unencrypted_",
+			Value: sops.DefaultUnencryptedSuffix,
+		},
+		cli.StringFlag{
+			Name:  "config",
+			Usage: "path to sops' config file. If set, sops will not search for the config file recursively.",
+		},
+	}
+	app.Action = func(c *cli.Context) error {
+		if c.NArg() < 1 {
+			return cli.NewExitError("Error: no file specified", 1)
+		}
+		file := c.Args()[0]
+		if _, err := os.Stat(file); os.IsNotExist(err) {
+			if c.String("add-kms") != "" || c.String("add-pgp") != "" || c.String("rm-kms") != "" || c.String("rm-pgp") != "" {
+				return cli.NewExitError("Error: cannot add or remove keys on non-existent files, use `--kms` and `--pgp` instead.", 49)
+			}
+			if c.Bool("encrypt") || c.Bool("decrypt") || c.Bool("rotate") {
+				return cli.NewExitError("Error: cannot operate on non-existent file", 100)
+			}
+		}
+		fileBytes, err := ioutil.ReadFile(file)
+
+		var output *os.File
+		if c.Bool("in-place") {
+			var err error
+			output, err = os.Create(file)
+			if err != nil {
+				return cli.NewExitError(fmt.Sprintf("Could not open in-place file for writing: %s", err), 10)
+			}
+			defer output.Close()
+		} else {
+			output = os.Stdout
+		}
+		if err != nil {
+			return cli.NewExitError("Error: could not read file", 2)
+		}
+		if c.Bool("encrypt") {
+			return encrypt(c, file, fileBytes, output)
+		} else if c.Bool("decrypt") {
+			return decrypt(c, file, fileBytes, output)
+		} else if c.Bool("rotate") {
+			return rotate(c, file, fileBytes, output)
+		} else {
+
+		}
+		return nil
+	}
+	app.Run(os.Args)
+}
+
+func runEditor(path string) {
+	editor := os.Getenv("EDITOR")
+	var cmd *exec.Cmd
+	if editor == "" {
+		cmd := exec.Command("which", "vim", "nano")
+		out, err := cmd.Output()
+		if err != nil {
+			panic("Could not find any editors")
+		}
+		cmd = exec.Command(strings.Split(string(out), "\n")[0], path)
+	} else {
+		cmd = exec.Command(editor, path)
+	}
+	cmd.Run()
+}
+
+func store(path string) sops.Store {
+	if strings.HasSuffix(path, ".yaml") {
+		return &yaml.YAMLStore{}
+	} else if strings.HasSuffix(path, ".json") {
+		// return &json.JSONStore{}
+	}
+	panic("Unknown file type for file " + path)
+}
+
+func findKey(keysources []sops.KeySource) (string, error) {
+	for _, ks := range keysources {
+		for _, k := range ks.Keys {
+			key, err := k.Decrypt()
+			if err == nil {
+				return key, nil
+			}
+		}
+	}
+	return "", fmt.Errorf("Could not get master key")
+}
+
+func decrypt(c *cli.Context, file string, fileBytes []byte, output io.Writer) error {
+	store := store(file)
+	metadata, err := store.LoadMetadata(string(fileBytes))
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error loading file: %s", err), 3)
+	}
+	key, err := findKey(metadata.KeySources)
```

Comment:
The type is `KeySource`, so I'd still have to iterate over each one and call GetDataKey.